### PR TITLE
Revert "Merge pull request #20124 from umanwizard/backport_57.9_reverts"

### PR DESCRIPTION
**Pull request by antiguru**
### Motivation

Second attempt at implementing feature #19348. The first (#18766) had a defect that it would keep arrangements around for operators that would otherwise drop their trace. This implementation should not do this anymore by only keeping a weak trace handle instead of a strong handle.

### Tips for reviewer

Best viewed commit-by-commit:
* Revert of the revert
* Removing some things from the revert that we didn't actually need.
* Update documentation
* Use a weak handle instead of a trace, adjust internal views
* Introduce the `enable_arrangement_size_logging` feature flag.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by antiguru** [comment]
Feature benchmark: https://buildkite.com/materialize/nightlies/builds/2994

File: src/sql/src/session/vars.rs [hunk 1]
```diff
@@ -1257,26 +1257,35 @@ mod cluster_scheduling {
 ///
 /// Note that not all `ServerVar<bool>` are feature flags. Feature flags are for variables that:
 /// - Belong to `SystemVars`, _not_ `SessionVars`
-/// - Default to false and must be explicitly enabled
+/// - Default to false and must be explicitly enabled or provide an optional value and internal flag.
```
Comment (jkosh44) [inline src/sql/src/session/vars.rs:1260]: I had trouble understanding this sentence without reading the actual code. Is it possible to add a doc-comment to each variant? Or maybe spell out all variants and their defaults here?

**Review by jkosh44 — APPROVED** [review]
Adapter code LGTM.

File: src/sql/src/session/vars.rs (hunk 1 above)
Comment (antiguru) [reply src/sql/src/session/vars.rs]: Yep, updated the underlying PR. Thanks!

File: src/compute-client/src/protocol/command.rs [hunk 2]
```diff
@@ -339,6 +339,8 @@ pub struct ComputeParameters {
     pub tracing: TracingParameters,
     /// gRPC client configuration.
     pub grpc_client: GrpcClientParameters,
+    /// Enable arrangement size logging
+    pub enable_arrangement_size_logging: Option<bool>,
```
Comment (teskje) [inline src/compute-client/src/protocol/command.rs:343]: Nit: Can you move that up to `enable_mz_join_core`, to keep the feature flags close together?

File: src/compute/src/logging/compute.rs [hunk 3]
```diff
@@ -15,6 +15,7 @@ use std::collections::{BTreeMap, BTreeSet, VecDeque};
 use std::rc::Rc;
 use std::time::Duration;
 
+use crate::compute_state::ComputeState;
```
Comment (teskje) [inline src/compute/src/logging/compute.rs:18]: Nit: import block

File: src/compute/src/logging/differential.rs [hunk 4]
```diff
@@ -15,6 +15,7 @@ use std::collections::BTreeMap;
 use std::rc::Rc;
 use std::time::Duration;
 
+use crate::compute_state::ComputeState;
```
Comment (teskje) [inline src/compute/src/logging/differential.rs:18]: Nit: import block

File: src/compute/src/logging/reachability.rs [hunk 5]
```diff
@@ -14,6 +14,7 @@ use std::collections::BTreeMap;
 use std::convert::TryInto;
 use std::rc::Rc;
 
+use crate::compute_state::ComputeState;
```
Comment (teskje) [inline src/compute/src/logging/reachability.rs:17]: Nit: import block

File: src/compute/src/logging/timely.rs [hunk 6]
```diff
@@ -15,6 +15,7 @@ use std::collections::BTreeMap;
 use std::rc::Rc;
 use std::time::Duration;
 
+use crate::compute_state::ComputeState;
```
Comment (teskje) [inline src/compute/src/logging/timely.rs:18]: Nit: import block

**Review by teskje — CHANGES_REQUESTED** [review]
Great to see this back on the way! I'm so happy that we didn't have to go with the `unsafe` solution. I added a couple of comments about adjustments to the introspection views we should do. The rest is nits.

Do we have an issue for keeping track of removing the feature flag after some time? It's necessary now, but we'll want to get rid of all the noise it introduces once it isn't anymore.

**Comment by antiguru** [comment]
#21131 to track removing the feature flag.

File: doc/user/content/sql/system-catalog/mz_internal.md [hunk 7]
```diff
@@ -959,14 +965,15 @@ The `mz_peek_durations_histogram` view describes a histogram of the duration in
 The `mz_records_per_dataflow` view describes the number of records in each [dataflow].
```
[the remark this answers is not in the document]
Comment (antiguru) [reply doc/user/content/sql/system-catalog/mz_internal.md:965]: I think we should rename it, but I'd rather not do it as part of this PR. It'll need sequencing with the console I believe, and it'll certainly take a few iterations to fix all the tests. I field #21132 to capture this.

File: src/adapter/src/catalog/builtin.rs [hunk 8]
```diff
@@ -2412,16 +2412,16 @@ SELECT
     dod.name,
     dod.worker_id,
     dod.dataflow_id,
+    ar_size.batches,
     ar_size.records,
     ar_size.size,
     ar_size.capacity,
     ar_size.allocations
```
[the remark this answers is not in the document]
Comment (antiguru) [reply src/adapter/src/catalog/builtin.rs:2419]: I was wondering the same, but I decided not to. The reason is that there is a different meaning of `null` vs. 0, where `null` indicates absence and `0` means there could be data, but there is none at the moment. The higher-level `mz_records_per_dataflow_operator` et al. views coalesce to 0.

[the remark this answers is not in the document]
Comment (teskje) [reply src/adapter/src/catalog/builtin.rs:2419]: So far all the non-`_per_worker` relations have the same schema as the `_per_worker` relations, except that the per-worker values are aggregated. This is how we get away with not explicitly documenting the `_per_worker` relations in the docs. I wouldn't want to deviate from this pattern here, so if you feel strongly about keeping the NULLs here, I'd rather keep them in `mz_records_per_dataflow_operator` as well.

File: doc/user/content/sql/system-catalog/mz_internal.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (teskje) [reply doc/user/content/sql/system-catalog/mz_internal.md:965]: SGTM!

File: src/adapter/src/catalog/builtin.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (antiguru) [reply src/adapter/src/catalog/builtin.rs]: Ok, this makes sense! I changed the PR to use `COALESCE` on all left outer joins to replace potential null values with 0. We could revisit this in future, but I think it's outside the scope of this PR to come to a conclusion on this topic. Thank you!

File: src/adapter/src/catalog/builtin.rs [hunk 9]
```diff
@@ -3227,16 +3230,16 @@ heap_allocations_cte AS (
 SELECT
     batches_cte.operator_id,
     batches_cte.worker_id,
-    records_cte.records,
     batches_cte.batches,
+    records_cte.records,
     heap_size_cte.size,
     heap_capacity_cte.capacity,
     heap_allocations_cte.allocations
```
[the remark this answers is not in the document]
Comment (antiguru) [reply src/adapter/src/catalog/builtin.rs]: Yup, added a `COALESCE`.

**Review by teskje — APPROVED** [review]
Thank you!

**Comment by antiguru** [comment]
Thanks for the reviews!

**Pull request merged**